syntax = "proto3";

package badge;

// badge service definition
service BadgeService {
    // migrate mongo to sql 
    rpc MigrateDataMongoToSQL(MigrationRequestMongoToSQL) returns (MigrationReply) {}

    // migrate mongo to mongo
    rpc MigrateDataMongo(MigrationRequestMongo) returns (MigrationReply) {}

    // migrate sql to sql
    rpc MigrateDataSQL(MigrationRequestSQL) returns (MigrationReply){}

    // migrate sql to mongo
    rpc MigrateDataSQLToMongo(MigrationRequestSQLToMongo) returns (MigrationReply){}

    // get all badges as a list
    rpc GetBadges(BadgeRequest) returns (Badges) {}
    // get all badges as a list from mysql table
    rpc GetBadgesMysql(Connection) returns (Badges) {}

    // get all badges in a stream
    rpc GetStreamBadge(BadgeRequest) returns (stream Badge){}
    // get all badges in a stream from mysql table
    rpc GetStreamBadgeMysql(Connection) returns (stream Badge){}
    
}

// will be used to coordinate with the server what the user wants
message BadgeRequest{
    string name = 1;
}
message MigrationRequestMongoToSQL{
    ConnectionIOMongo origin=1;
    ConnectionIOSQL destination=2;
    string table = 3;
    string type = 8;
}
message MigrationRequestSQL{
    ConnectionIOSQL origin=1;
    ConnectionIOSQL destination=2;
    string table = 3;
    string type = 8;
}
message MigrationRequestMongo{
    ConnectionIOMongo origin=1;
    ConnectionIOMongo destination=2;
    string table = 3;
    string type = 8;
}
message MigrationRequestSQLToMongo{
    ConnectionIOSQL origin=1;
    ConnectionIOMongo destination=2;
    string table = 3;
    string type = 8;
}
message MigrationReply{
    string outcome =1;
}
message Connection{
    optional string connection =1;
    string username = 2;
    string password = 3;
    string host = 4;
    optional int32 port = 5;
    string database = 6;
    string table = 7;
}
message ConnectionIOSQL{
    string user = 2;
    string password = 3;
    string host = 4;
    int32 port = 5;
    string database = 6;
    repeated int32 client_flags = 9;
    optional string ssl_ca = 10;
}


message ConnectionIOMongo{
    string connectionString = 1;
    string database = 2;
}

message ConnectionWithMetadata{
    string connection =1;
    string username = 2;
    string password = 3;
    string host = 4;
    int32 port = 5;
    string database = 6;
    string tableIn = 7;
    string tableOut = 8;
    repeated string rows = 9;
    repeated string types = 10;
    string primaryKey = 11;
}


// list of Badges
message Badges {
    repeated Badge badges = 1;
}


// single badge
// follows the database keys in badges
message Badge{
    int64 Id = 1;
    int64 UserId = 2;
    string Name = 3;
    string Date = 4;
    int64 Class = 5;
    string TagBased = 6;
    
}